README.md

```diff
@@ -183,6 +183,22 @@ Where `fortigate-key.yaml` contains pairs of FortiGate targets and API keys in t
 NOTE: Currently only token authentication is supported. FortiGate does not allow usage of tokens on non-HTTPS connections,
 which means that currently you need HTTPS to be configured properly.
 
+You can select which probes or probe categories you want to run per target, for example:
+
+```
+"https://my-fortigate":
+  token: api-key-goes-here
+  probes:
+    - System
```

Comment:
Since it is not clear if `System` is a probe or a category, maybe we should do some kind of prefix tier system?

Like you define your metric as `/BGP/BGPNeighborsIPv4` - and that means that somebody can do either:
```yaml
probes:
  - /BGP
```
or
```yaml
probes:
  - /BGP/BGPNeighborsIPv4
```

Code-wise we could just use something like `strings.HasPrefix`.

Reply:
I am OK with your tier modification.
This is actually more convenient and flexible on long term with an increasing number of probes.
I think the best solution is to support both :
- probe unique full name (ex: `BGPNeighborsIPv4`) - must be equal to current `aProbe.name`
- probe unique category full name (ex: `BGP/Neighbors/IPv4`) - must `HasPrefix` with `aProbe.category`

Then `BGP/Neighbors/IPv4` and `BGPNeighborsIPv4` are equivalent and can both be used to identify a unique probe.

For prefix matching, HasPrefix offers flexible solution for, multi tier match at any level.
Example :
```
- BGP                   (match all BGP probes)
- BGP/                  (match all BGP probes)
- BGP/Neighb            (match all 4 BGP probes)
- BGP/Neighbors         (match 2 BGP probes BGPNeighborsIPv4 and BGPNeighborsIPv6)
- BGP/Neighbors/        (match 2 BGP probes BGPNeighborsIPv4 and BGPNeighborsIPv6)
- BGP/Neighbors/IPv     (match 2 BGP probes BGPNeighborsIPv4 and BGPNeighborsIPv6)
- BGP/Neighbors/IPv4    (match 1 BGP probes BGPNeighborsIPv4)
- BGP/Neighbors/IPv44   (match nothing)
```
This solution looks quite optimized and polyvalent .


One step forward, we can obtain `BGPNeightborsIPv4` from `BGP/Neighbors/IPv4` by just removing slashes.
This simplifies the current struct to:
```golang
type probeDetailedFunc struct {
        name string
	function probeFunc
}
````
`name` is now storing former category (ex: `BGP/Neighbors/IPv4`).
`probe.go` declaration section would be more simple.

Do you think this is a good idea to go with the slashes removing, instead of storing both category and name?

Reply:
I really like the idea of this prefix system, but I propose a slight modification. 

### All probes are executed
``` yaml
"https://my-fortigate":
  token: api-key-goes-here
    # or entirely dropped -> all probes are used
``` 
``` yaml
"https://my-fortigate":
  token: api-key-goes-here
  probes:  # or entirely dropped -> all probes are used
``` 
``` yaml
"https://my-fortigate":
  token: api-key-goes-here
  probes:
    include: [ ] # -> all probes are used
```
### Only matching probes are executed
``` yaml
"https://my-fortigate":
  token: api-key-goes-here
  probes:
    include: 
      - System
```


This allows to add additional options later on such like this and is more readable: 
``` yaml
"https://my-fortigate":
  token: api-key-goes-here
  probes:
    include: 
      - System
    exclude: 
      - System/SystemHAStatistics
```


Regarding the slash replacement.    
I'm not sure what you mean with 'slash removing' as I understand your latest proposal would be to store in the style of : 
```
[
  { "System/SystemHAStatistics", probeSystemHAStatistics },
...
]
```

Reply:
By 'slash removing' I mean we can just store in `probe.go`
```
for _, aProbe := range []probeDetailedFunc{
		{"BGP/Neighbors/IPv4", probeBGPNeighborsIPv4},
                {"System/HAStatistics", probeSystemHAStatistics},
                ...
```
User can either add `BGP/Neighbors/IPv4` or `BGPNeighborsIPv4` which is totally equivalent for including this probe.

### Exclusion with "!"
I was also thinking to an exclude capability as a next step, but as you talked about it we can plan to add it all at once.
Your proposal looks fine to me.

I was more thinking to a "!" syntax.
Processing of inclusion, exclusion could be done sequentially.
The logic :
- if list start with an include statement, then initialize all other probes to "not wanted"
- if list start with an exclude statement, then initialize all probes to "wanted", expect those matching the exclude statement
- for 2nd to N term of the list, just include or exclude matching probes for each list entry

Summing up and detailing use cases :
```yaml
probes: System
# select only all System prefixed probes

probes: BGP/Neighbors/IPv4
# select only BGP/Neighbors/IPv4

probes: BGPNeighborsIPv4
# equivalent to previous declaration

probes: BGPNeighborsIPv4, BGP/Neighbors/IPv6
# mixing inclusion system category/name to include both BGPNeighborsIPv4 AND BGPNeighborsIPv6

probes: System, BGP
# select all System AND BGP prefixed probes

probes: !System
# select all probes but those prefixed by System

probes: !System, BGP
# not really clever but select all probes but those prefixed by System

probes: BGP, !System
# not really clever but just select all BGP prefixed probes

probes: !BGP, BGP/Neighbors
# select all probes but those prefixed by BGP + the 2 probes starting with BGP/Neighbors

probes: BGP/Neighbors, !BGP
# actually select no probe, because first select 2 BGP probes, excluded by !BGP

probes: BGP/Neighbors, !BGP, BGP/Neighbors
# not really clever but just select the 2 probes starting with BGP/Neighbors, then exclude all BGP probes, and finally select the 2 probes starting with BGP/Neighbors again
```

pros :
- one list declaration for include/exclude
- possible do declare a complex inclusion exclusion sequence (order matters)
- go through 1 list instead of 2

cons :
- require to add a check on "!" prefix for probe name
- maybe a bit less simple to read

What do you think about this inclusion/exclusion method ?

### Probe name constraints
I don't know if there is already a naming convention for probes.
I propose the following probe name constraints to make probe naming more strait forward from Fortigate API.
- full name ex `SystemHAstatistics` is obtained from categorized name ex: `System/HAstatistics` which is stored in `pkg/probe/probe.go`
- probe name cannot start with "!"
- probe full name should respect regex : `^[a-ZA-Z0-9]{4,}$`
- probe categorized name must respect regex : `^([a-ZA-Z0-9]+/)+[a-ZA-Z0-9]+$`
- probe name should follow the fortigate API syntax ex: `System/HAstatistics` for `/api/v2/monitor/system/ha-statistics`
- probe full name and categorized name must be unique across all probes

Right now, not all probes respect the entire naming convention, for exemple:
`/api/v2/monitor/router/bgp/neighbors` should be `router/BGP/Neighbors/IPv4` and `routerBGPNeighborsIPv4`

What do you think of those probe name constraints ? Do you think applying this to the entire project is the right thing to do ?

Reply:
I am with @secustor here - I like include/exclude better than `!` for the simple fact that it is less string-mangling. We have a structured configuration file format already (YAML), we are probably better off using it to communicate meaning as much as possible.

I am a fan of having less ways of doing the same thing, so if we can agree on that the probe name is for example `BGP/Neighbors/IPv4` then I would prefer that be the only way it would be written. I think introducing multiple ways of referring to a single probe adds a lot of confusion for a small benefit for some people that do not like the `/` character.

Finally - I am all for your naming rules @IMSdevsecu, we should add them to the contribution guide (which isn't written yet, https://github.com/bluecmd/fortigate_exporter/issues/64). The only part I want to relax is the "should follow Fortigate API" bit. Why? Many metrics requires fetting both from monitor and potentially cmdb which both [can vary](https://github.com/bluecmd/fortigate_exporter/blob/e4274022852cfcc6f196180bf88fa3ba3d57f11d/pkg/probe/system_ha_statistics.go#L86-L98) in how they look like - besides, we probably don't want to break people's configs just because Fortinet changed their APIs between e.g. 6.x and 7.x.

Instead, I would formulate your "probe name should follow the fortigate API syntax" as a recommendation as to what to name the probe function when it is first introduced, with two notable exceptions:
- If there already exists similar metrics, stick to that naming convention
- Shorter is better (E.g. `Router/BGP/x` is redundant, use `BGP/x`)

Also, @secustor - another example that follows logically given that we are prefix matching:
```yaml
"https://my-fortigate":
  token: api-key-goes-here
  probes:
    include: 
       - '' # -> all probes are used
```

Reply:
> I am with @secustor here - I like include/exclude better than `!` for the simple fact that it is less string-mangling. We have a structured configuration file format already (YAML), we are probably better off using it to communicate meaning as much as possible.

I understand. This is more simple to code, use and document. This also allows for less complex selection.
If we go for 2 distinct lists which one to process first ?
Because default is "all probe wanted", I think we can process this way:
```
foreach probe 
    if "include" list is not empty then
        default to "not wanted"
        if probe as Prefix in "include" list then
            set to "wanted"
    else
        default to "wanted"

    if "exclude" list is not empty then
         if probe as Prefix in "exclude" list then
            set to "not wanted"
```
Is this algorithm okay for you ?

Reply:
> I am a fan of having less ways of doing the same thing, so if we can agree on that the probe name is for example `BGP/Neighbors/IPv4` then I would prefer that be the only way it would be written. I think introducing multiple ways of referring to a single probe adds a lot of confusion for a small benefit for some people that do not like the `/` character.

Fine let's keep the `BGP/Neighbors/IPv4` name only.

Reply:
> Finally - I am all for your naming rules @IMSdevsecu, we should add them to the contribution guide (which isn't written yet, #64). The only part I want to relax is the "should follow Fortigate API" bit. Why? Many metrics requires fetting both from monitor and potentially cmdb which both [can vary](https://github.com/bluecmd/fortigate_exporter/blob/e4274022852cfcc6f196180bf88fa3ba3d57f11d/pkg/probe/system_ha_statistics.go#L86-L98) in how they look like - besides, we probably don't want to break people's configs just because Fortinet changed their APIs between e.g. 6.x and 7.x.
> 
> Instead, I would formulate your "probe name should follow the fortigate API syntax" as a recommendation as to what to name the probe function when it is first introduced, with two notable exceptions:
> 
>     * If there already exists similar metrics, stick to that naming convention
> 
>     * Shorter is better (E.g. `Router/BGP/x` is redundant, use `BGP/x`)
> 
> 
> Also, @secustor - another example that follows logically given that we are prefix matching:
> 
> ```yaml
> "https://my-fortigate":
>   token: api-key-goes-here
>   probes:
>     include: 
>        - '' # -> all probes are used
> ```

Sounds good to me.
You must have a deeper knowledge of the evolution of their API. I trust you if you prefer to keep some flexibility with the naming 😉